Fix/Improve `processWindowErrorEvent`

silverwind opened the pull request:
- `e.error` can be undefined in some cases which would raise an error inside this error handler, fixed that.
- The displayed message mentions looking into the console, but in my case of error from `ResizeObserver` there was nothing there, so add this logging. I think this logging was once there but got lost during refactoring.

silverwind commented:
Maybe browser does the logging, but only if there is `e.error`, need to test.

silverwind commented:
Putting draft until I have confirmed above behaviour. Definitely should backport this, otherwise some "errors" will cause errors in inside this error handler.

silverwind commented:
Rewrote the function, now it truly handles all cases and the comments accurately describe what it does.

Flashes:
<img width="828" alt="Screenshot 2024-02-26 at 20 14 57" src="https://github.com/go-gitea/gitea/assets/115237/74108576-c337-4184-84ba-5dbdd1cc01c2">

Console.
<img width="815" alt="Screenshot 2024-02-26 at 20 15 04" src="https://github.com/go-gitea/gitea/assets/115237/14e703fe-4cc7-44e9-870d-58a5188f6ec3">

silverwind commented on web_src/js/bootstrap.js:
```diff
@@ -28,27 +28,32 @@ export function showGlobalErrorMessage(msg) {
-  showGlobalErrorMessage(`${message} Open browser console to see more details.`);
+  const renderedType = type === 'unhandledrejection' ? 'promise rejection' : type;
+  let msg = err?.message ?? message;
+  if (lineno) msg += `(${filename} @ ${lineno}:${colno})`;
```
```suggestion
  if (lineno) msg += ` (${filename} @ ${lineno}:${colno})`;
```

silverwind commented on web_src/js/bootstrap.js:
```diff
@@ -28,27 +28,32 @@ export function showGlobalErrorMessage(msg) {
-    message = `JavaScript error: ${e.message} (${e.filename} @ ${e.lineno}:${e.colno}).`;
+  // This handler not only receives errors but also error events, detectable by `err` being null
+  // or undefined. Error events do not log to the browser console by default but as they might
+  // still be relevant, we show them during development. In production they will only log to the
+  // console. References and examples:
```
```suggestion
  // still be relevant, we show them during development. References and examples:
```

silverwind commented on web_src/js/bootstrap.js:
```diff
@@ -28,27 +29,31 @@ export function showGlobalErrorMessage(msg) {
-    message = `JavaScript promise rejection: ${err.message}.`;
-  } else {
-    message = `JavaScript error: ${e.message} (${e.filename} @ ${e.lineno}:${e.colno}).`;
+  // This handler not only receives `Errors`` but also error events, detectable by `error` being
```
```suggestion
  // This handler not only receives `Error`s but also error events, detectable by `error` being
```

wxiaoguang reviewed (commented):
I don't like this change. Because "e" could be anything by design.

But I won't block it.

silverwind commented:
> I don't like this change. Because "e" could be anything by design.
> 
> But I won't block it.

I know you don't like such destructuring, but [ErrorEvent](https://developer.mozilla.org/en-US/docs/Web/API/ErrorEvent) and its properties are well defined, so I see no issue.

silverwind commented:
The handler also receives [PromiseRejectionEvent](https://developer.mozilla.org/en-US/docs/Web/API/PromiseRejectionEvent) and I see its `reason` property is not guaranteed to be an `Error`, so it could be improved to handle that case.

wxiaoguang commented:
> I know you don't like such destructuring

Not really. The real problem is `Because "e" could be anything by design.`. Before the code was quite straightforward: if this error, do this, if that error, do that. But the new code just mixed everything together. I don't like making readers "guess" what the type/value there should be.

wxiaoguang commented:
ps: the last time I said "no" to such syntax, the reason is the same: it make code unclear (https://github.com/go-gitea/gitea/pull/25253#discussion_r1229359091), and introduced new bugs https://github.com/go-gitea/gitea/pull/25253#discussion_r1230361844

silverwind commented:
I can restore the previous `type` check, no problem.

silverwind commented:
I decided against `type` check as there is no overlap in properties between ErrorEvent and PromiseRejectionEvent, so it would unnecessarily bloat the code. I fixed the jsDoc, added a typecheck for `Error` before accessing `stack`, and tweaked the comment again.

Everything works as expected:

<img width="1303" alt="Screenshot 2024-02-27 at 21 18 52" src="https://github.com/go-gitea/gitea/assets/115237/90870fa2-8518-439b-9118-cf8ae591beb5">

silverwind commented:
> Before the old code (before https://github.com/go-gitea/gitea/pull/29303) was quite straightforward: if this error, do this, if that error, do that. But the new code just mixed everything together. I don't like making readers "guess" what the type/value there should be.

It's pretty straightforward already. Could be made even more so by splitting into two handler functions hat converge into a logging function, but I think this be in overengineering territory. I don't see the current function as hard to read just because of this `error ?? reason`, both serve the same purpose.

silverwind commented:
@delvh @yardenshoham maybe you want to give this another look as it has changed a bit since your reviews. If there no further objections, I will merge this soon.

silverwind commented on web_src/js/bootstrap.js:
```diff
@@ -26,29 +27,42 @@ export function showGlobalErrorMessage(msg) {
 
 /**
- * @param {ErrorEvent} e
+ * @param {ErrorEvent|PromiseRejectionEvent} event- Event
```
```suggestion
 * @param {ErrorEvent|PromiseRejectionEvent} event - Event
```

yardenshoham commented:
Please recommend ways to test

silverwind commented:
Can be tested by installing Violentmonkey with such userscripts:

```js
// ==UserScript==
// @name        Gitea error trigger 1
// @match       http://localhost:3000/*
// ==/UserScript==
setTimeout(() => {
  throw new Error("sync error from userscript");
}, 10000)

const divElem = document.querySelector(".repo-header");
const resizeObserver = new ResizeObserver((entries) => {
  for (const entry of entries) {
    entry.target.style.width = entry.contentBoxSize[0].inlineSize + 10 + "px";
  }
});
resizeObserver.observe(divElem);
```
```js
// ==UserScript==
// @name        Gitea error trigger 2
// @match       http://localhost:3000/*
// ==/UserScript==
(async () => {
  throw new Error("async error from userscript");
})()
```

yardenshoham commented:
I can confirm I see the errors
![image](https://github.com/go-gitea/gitea/assets/20454870/f6bfe7f5-1b7a-4f56-aefd-d240b0874594)
![image](https://github.com/go-gitea/gitea/assets/20454870/0ced40fb-bec8-4f4c-b99b-5294f99d3ad4)

silverwind commented:
If you go on a repo page, you should also see the ResizeObserver error :)
